### ISSUE TITLE
How to use and debug mixed-precision in 1.6.0 ?

### ISSUE THREAD

**Issue by djordjemiladinovic** [pytorch/pytorch#42996]
## ❓ Questions and Help

Hi. I've been trying to take advantage of new mixed-precision set of tools, mainly following the instructions given in https://pytorch.org/docs/stable/amp.html#autocasting. My code is running, I see that the gradients are being scaled as expected, however, my memory footprint is de facto the same (inspected by nvidia-smi), and the overall execution time is even longer. I was wondering how I can debug this and what could possibly go wrong.

The changes to my code are minimal: (1) I added autocast context to the body of my forward method; and (2) gradient scaling updates as suggested in the tutorial. The operations in my code are standard, lots of CNN and torch.mm executions. My program is being executed on 8 GPUs, for data parallelism.

## Environment

 - PyTorch Version: 1.6.0
 - OS: Linux
 - How you installed PyTorch: pip
 - Python version: 3.7
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: 8 x Tesla V100

cc @mcarilli @jlin27

**Comment by dlrac** [comment]
You should check if Tensor cores are used at all, there are some dimension requirements on batch size and number of hidden units. With apex it was done using for example: https://github.com/NVIDIA/apex/tree/master/apex/pyprof. 

Btw, while being advertised,  mixed precision is not quite there yet, e.g. RNN modules/cells don't work at all see:

1. https://github.com/pytorch/pytorch/issues/42605
2. https://github.com/pytorch/pytorch/issues/36428

Also you mentioned data parallelism and I'm not sure to which parallelism model are you referring to, I think it doesn't work with DataParallel, it only works with DDP (I'm not 100% sure for this, there were some problems, maybe it is fixed now).

**Comment by djordjemiladinovic** [comment]
Thanks for the fast response !

I will try to see if Tensor cores are activated, it is not trivial to use the profiler since I am running my networks on a cluster. Will let you know if I discover something.

However in any case, it is a bit strange not see any difference at all, not even small memory reduction.

I do not get the error related to RNN modules so I do not think this applies to me.

I am using DDP indeed, however I even tried mixed-precision vs normal on a single GPU, still I can't see any difference.

**Comment by gchanan** [comment]
sounds like we may need better debugging docs.

**Comment by dlrac** [comment]
@djordjemiladinovic I guess the bottleneck for your model is memory and that is why you are resorting to mixed precision. If I'm right you should check out gradient checkpoint as well, that can save you memory in trade for computation (https://pytorch.org/docs/stable/checkpoint.html). However, I haven't had any luck using DDP and gradient checkpoint together, this technique only helps with single GPU in my knowledge (hope somebody will correct me here and provide the link to minimal working example).

**Comment by djordjemiladinovic** [comment]
@dlrac My model is rather large and I am trying to optimize both memory and execution time. To that end, I was really looking forward to AMP, and according to the [blog post](https://pytorch.org/blog/accelerating-training-on-nvidia-gpus-with-pytorch-automatic-mixed-precision/) I am supposed to get both at the same time. However, I was unable to make it work, not even after few days of testing..

Gradient checkpointing might be a solution in certain situations, however runtime is really critical for me :/ Thanks for the suggestion in any case !

**Comment by mcarilli** [comment]
> I think it doesn't work with DataParallel, it only works with DDP

`torch.cuda.amp` should work with either DP or DDP.
In the recommended (typically most performant) mode of DDP with one GPU per process, no changes are required to use Amp beyond what's documented for single-GPU usage.  
For DP or DDP with >1 GPU per process, [a minor change to model.forward is required](https://pytorch.org/docs/stable/notes/amp_examples.html#working-with-multiple-gpus).

However, it doesn't sound like multi-GPU is the issue.  I'd focus on single-GPU performance first.  A large model with CNNs and gemms should be an ideal amp target.  Are you sure you're saturating the device?  Also, for the gemms, are batch+input feature+output feature sizes a multiple of 8?

**Comment by djordjemiladinovic** [comment]
Hi @mcarilli I am now debugging in single-GPU mode, a model which is indeed a chain of intertwined convolutions and gemms, so I too think it is a primary example of a code to benefit from amp.

My batch sizes are 16, 32, 64, per GPU. Inputs are images of size 32x32x3.The (batch+input) and (feature+output feature) are however not necessarily multiples of 8. I was expecting to see some improvement nevertheless, but maybe now first I will try to satisfy these constraints too.

Please let me know if you have some hints in addition.

**Comment by djordjemiladinovic** [comment]
It seems I get no improvements even if all dimensions are multiples of 8.

**Comment by djordjemiladinovic** [comment]
Hi @mcarilli 

This is the example I created for you:

`
    import torch 
    import torch.nn as nn
    from torch.cuda.amp import autocast
    
    
    class LSTMCell(nn.Module):
        def __init__(self, input_size, hidden_size):
            super(LSTMCell, self).__init__()
            self.input_size = input_size
            self.hidden_size = hidden_size
            self.weight_ih = nn.Parameter(torch.randn(4 * hidden_size, input_size))
            self.weight_hh = nn.Parameter(torch.randn(4 * hidden_size, hidden_size))
            self.bias_ih = nn.Parameter(torch.randn(4 * hidden_size))
            self.bias_hh = nn.Parameter(torch.randn(4 * hidden_size))
    
        @autocast()
        def forward(self, input, state):
            # type: (Tensor, Tuple[Tensor, Tensor]) -> Tuple[Tensor, Tuple[Tensor, Tensor]]
            hx, cx = state
            gates = (torch.mm(input, self.weight_ih.t()) + self.bias_ih +
                     torch.mm(hx, self.weight_hh.t()) + self.bias_hh)
            ingate, forgetgate, cellgate, outgate = gates.chunk(4, 1)
    
            # debug
            print("ingate type", ingate.type())
            print("ingate half type", ingate.half().type())
            print("Autocast enabled?", torch.is_autocast_enabled())
    
    
    
            ingate = torch.sigmoid(ingate)
            forgetgate = torch.sigmoid(forgetgate)
            cellgate = torch.tanh(cellgate)
            outgate = torch.sigmoid(outgate)
    
            cy = (forgetgate * cx) + (ingate * cellgate)
            hy = outgate * torch.tanh(cy)
    
            return hy, (hy, cy)
    
    
    if __name__ == '__main__':
    
        # detect device
        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    
        # intialize
        lstm = LSTMCell(100, 100).to(device)
        input_tensor = torch.randn(32, 100, device=device)
        h = torch.randn(32, 100, device=device)
        c = torch.randn(32, 100, device=device)
    
        # check device just in case
        print("device: ", h.device)
    
        # forward
        h, c = lstm(input_tensor, (h, c))
    
        print("Success")
`

When I run it on Tesla V100 i get:

device:  cuda:0
ingate type torch.cuda.FloatTensor
ingate half type torch.cuda.HalfTensor
Autocast enabled? True
Success

**Comment by djordjemiladinovic** [comment]
I was debugging myself a bit. It seems that the output of torch.mm(input, self.weight_ih.t()) is a HalfFloat, however the output of torch.mm(input, self.weight_ih.t()) + self.bias_ih is a Float, because of the "addition" operation.

Is this expected behaviour ? It still remains a question why the overall memory requirements are not decreased at all.

**Comment by mcarilli** [comment]
> the output of torch.mm(input, self.weight_ih.t()) + self.bias_ih is a Float, because of the "addition" operation.

This is expected type-promotion behavior if `self.bias_ih` is fp32.

Try replacing `torch.mm(input, self.weight_ih.t()) + self.bias_ih` with `torch.addmm` or `torch.nn.functional.linear`.  [`linear`'s source](https://pytorch.org/docs/stable/_modules/torch/nn/functional.html#linear) uses `addmm` in a way that looks equivalent to yours, and `addmm` should receive autocasting as a unified chunk, so there won't be problems with operator+'s implicit promotion.

Hmm i should add that tip to a best practices guide or something, if you can confirm it works

**Comment by djordjemiladinovic** [comment]
Hi @mcarilli 

I finally tested what you proposed i.e. using "addmm" instead of "mm" and "add" separately and it seems like it works, I get both memory and speed improvements. In this mini example, I get HalfTensor where I expect. Your suggestion helped, thanks ! I too think it would be a valuable tip for practitioners.

I am aware of LSTM and LSTMCell APIs, however in fact I am using a really customized units, so that does not work for me. I only used LSTM above as a an example.
